### PR TITLE
Search films

### PR DESCRIPTION
**Pull request by Falya**

### CODE REVIEW

File: src/methods/clientMethods.mjs [hunk 1]
```diff
@@ -5,36 +5,62 @@ import Seance from '../models/Seance.mjs';
...
+    console.error(error);
+  }
+}
+
 export async function getMovieSeances(params) {
   const { movieId, movieTheaterId, city, features, date } = params;
   const nowTime = new Date();
   const today = new Date(date);
 
+  const { emptyMoreOne, emptyMoreTwo, hasEmptyVip, hasEmptyDouble, video3d } = features;
```
Comment (dzmitry-penda) [inline src/methods/clientMethods.mjs]: if it's boolean, try to name it more like boolean

File: src/namedMessages/namedMessages.mjs [hunk 2]
```diff
@@ -19,6 +19,7 @@ const messages = {
   PAYMENT_FAILED: 'Failed payment',
   PAYMENT_TICKETS_SUCCESS: 'Tickets added to user',
   PAYMENT_TICKETS_FAILED: 'Tickets hasn`t been add',
+  USER_PROFILE_FAILED: 'Something was wrong',
```
Comment (dzmitry-penda) [inline src/namedMessages/namedMessages.mjs]: Not meaningful at all

File: src/utils/utils.mjs [hunk 3]
```diff
@@ -0,0 +1,52 @@
+export function filterSeancesByFeatures(seances, emptyAmount, hasEmptyVip, hasEmptyDouble) {
+  let filteredSeances = seances;
+
+  if (emptyAmount) {
```
Comment (dzmitry-penda) [inline src/utils/utils.mjs]: split it into separate methods, think of applying some kind of pattern here :)

File: src/methods/clientMethods.mjs (hunk 1 above)
Comment (Falya) [reply src/methods/clientMethods.mjs]: fixed

File: src/namedMessages/namedMessages.mjs (hunk 2 above)
Comment (Falya) [reply src/namedMessages/namedMessages.mjs]: fixed

File: src/utils/utils.mjs (hunk 3 above)
Comment (Falya) [reply src/utils/utils.mjs]: fixed

File: src/utils/utils.mjs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  hasEmptyVip && filterList.push(emptyVipFilter);
+  hasEmptyDouble && filterList.push(emptyDoubleFilter);
+
+  return filterList.reduce((acc, filter) => {
+    acc = filter(acc, emptyAmount);
+    return acc;
+  }, seances);
+}
+
+function emptyVipFilter(seances, emptyAmount) {
```
Comment (dzmitry-penda) [inline src/utils/utils.mjs]: emptyVipFilter and emptyDoubleFilter are basically the same
DRY

File: src/methods/clientMethods.mjs [hunk 5]
```diff
@@ -5,36 +5,62 @@ import Seance from '../models/Seance.mjs';
...
 
   today.setHours(23, 59);
+
   let theaters = [];
 
   try {
     const movieTheaterQuery = MovieTheater.find();
     const seancesQuery = Seance.find({ date: { $lte: today, $gte: nowTime }, movieName: movieId }).sort('date');
 
+    hasVideo3d && seancesQuery.where({ 'format.video': '3D' });
```
Comment (dzmitry-penda) [inline src/methods/clientMethods.mjs]: this should be also a filter

File: src/methods/clientMethods.mjs [hunk 6]
```diff
@@ -5,36 +5,62 @@ import Seance from '../models/Seance.mjs';
...
+}
+
 export async function getMovieSeances(params) {
   const { movieId, movieTheaterId, city, features, date } = params;
   const nowTime = new Date();
   const today = new Date(date);
 
+  const { hasEmptyMoreOne, hasEmptyMoreTwo, hasEmptyVip, hasEmptyDouble, hasVideo3d } = features;
+
+  const emptyAmount = hasEmptyMoreTwo ? 2 : hasEmptyMoreOne ? 1 : null;
```
Comment (dzmitry-penda) [inline src/methods/clientMethods.mjs]: this can be moved to filters file

File: src/methods/clientMethods.mjs [hunk 7]
```diff
@@ -246,53 +272,57 @@ export async function compareOrder(params) {
...
-    email: user.email,
-    registredAt: user.registredAt,
-    orders: grouped,
-    bought: user.tickets.length,
-    lastPurchase: user.tickets[user.tickets.length - 1].buyingTime,
-  };
+      newSeance.tickets = userInfo.tickets
+        .filter(ticket => ticket.seanceId._id === seance._id)
+        .map(({ buyingTime, price, rowNumber, seatNumber, seatType }) => {
+          return { buyingTime, price, rowNumber, seatNumber, seatType };
```
Comment (dzmitry-penda) [inline src/methods/clientMethods.mjs]: map(({ buyingTime, price, rowNumber, seatNumber, seatType }) => ({ buyingTime, price, rowNumber, seatNumber, seatType }))

File: src/utils/utils.mjs [hunk 8]
```diff
@@ -0,0 +1,54 @@
+export function filterSeancesByFeatures(seances, emptyAmount, hasEmptyVip, hasEmptyDouble) {
+  let filterList = [];
+  emptyAmount && filterList.push(emptySeatsFilter);
```
Comment (dzmitry-penda) [inline src/utils/utils.mjs]: can you take object, containing these params, take it keys and values - convert them to array, filter unused query parameters  out of resuting array and reduce it to form the complete query

File: src/utils/utils.mjs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      }, 0) - soldSeats.length;
+    return emptySeats > emptyAmount;
+  });
+}
+
+function emptySeatsFilter(seances, emptyAmount) {
+  return seances.filter(seance => {
+    const emptySeats =
+      seance.hallId.rows.reduce((sum, row) => {
+        sum += row.rowLength;
```
Comment (dzmitry-penda) [inline src/utils/utils.mjs]: seance.hallId.rows.reduce((sum, row) =>  sum + row.rowLength

File: src/utils/utils.mjs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+function emptyDoubleFilter(seances, emptyAmount) {
+  return seances.filter(seance => {
+    let soldSeats = [];
+    const emptySeats =
+      seance.hallId.rows.reduce((sum, row) => {
+        if (row.rowType === 'double') {
+          sum += row.rowLength;
+          const currentRow = seance.soldSeats.filter(({ rowNumber }) => rowNumber === row.rowNumber);
+          soldSeats = [...soldSeats, ...currentRow];
```
Comment (dzmitry-penda) [inline src/utils/utils.mjs]: reduce should not influence anything outside callback

File: src/utils/utils.mjs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    return acc;
+  }, seances);
+}
+
+function emptyVipFilter(seances, emptyAmount) {
+  return seances.filter(seance => {
+    let soldSeats = [];
+    const emptySeats =
+      seance.hallId.rows.reduce((sum, row) => {
+        if (row.rowType === 'vip') {
```
Comment (dzmitry-penda) [inline src/utils/utils.mjs]: this is a job for filter() method

File: src/utils/utils.mjs [hunk 12]
```diff
@@ -0,0 +1,54 @@
+export function filterSeancesByFeatures(seances, emptyAmount, hasEmptyVip, hasEmptyDouble) {
+  let filterList = [];
+  emptyAmount && filterList.push(emptySeatsFilter);
+  hasEmptyVip && filterList.push(emptyVipFilter);
+  hasEmptyDouble && filterList.push(emptyDoubleFilter);
+
+  return filterList.reduce((acc, filter) => {
```
Comment (dzmitry-penda) [inline src/utils/utils.mjs]: filterList.reduce((acc, filter) => filter(acc, emptyAmount), seances)

File: src/utils/utils.mjs (hunk 10 above)
Comment (Falya) [reply src/utils/utils.mjs]: fixed

File: src/utils/utils.mjs (hunk 9 above)
Comment (Falya) [reply src/utils/utils.mjs]: fixed

File: src/methods/clientMethods.mjs (hunk 7 above)
Comment (Falya) [reply src/methods/clientMethods.mjs]: fixed

File: src/utils/utils.mjs (hunk 12 above)
Comment (Falya) [reply src/utils/utils.mjs]: fixed

File: src/utils/utils.mjs (hunk 8 above)
Comment (Falya) [reply src/utils/utils.mjs]: Did something similar

File: src/utils/utils.mjs (hunk 11 above)
Comment (Falya) [reply src/utils/utils.mjs]: fixed

File: src/methods/clientMethods.mjs (hunk 6 above)
Comment (Falya) [reply src/methods/clientMethods.mjs]: fixed

File: src/methods/clientMethods.mjs (hunk 5 above)
Comment (Falya) [reply src/methods/clientMethods.mjs]: fixed

File: src/utils/utils.mjs (hunk 4 above)
Comment (Falya) [reply src/utils/utils.mjs]: fixed

**Pull request merged**